app/models/test_run.rb

```diff
@@ -109,6 +115,33 @@ def create_test_group_result(result)
     )
   end
 
+  def add_extra_marks(extra_marks)
+    extra_marks_total = 0
+    submission_result = self.grouping&.current_submission_used&.get_latest_result
+    return extra_marks_total if submission_result.nil?
+
+    extra_marks.each do |extra_mark|
+      unit = extra_mark['unit']
+      mark = extra_mark['mark']
+
+      unless ExtraMark.valid_unit?(unit)
```

Comment:
This part isn't necessary, as there's a validation on `ExtraMark` that will check the `unit`. We can use `ExtraMark.create` instead of `ExtraMark.create!`, which will not raise an error if there's an invalid unit. This is analogous to what we're doing with annotations (see `create_annotations`).